lib/events/s3_exporter.rb

```diff
@@ -0,0 +1,63 @@
+module Events
+  class S3Exporter
+    def initialize(created_before, created_on_or_after = nil)
+      @created_before = created_before
+      @created_on_or_after = created_on_or_after
+    end
+
+    def export
+      file = Tempfile.new("events_s3_#{filename}.csv.gz")
+      zipped = Zlib::GzipWriter.open(file) do |gzip|
+        gzip.orig_name = "#{filename}.csv"
+        build_csv(gzip)
+        gzip.close
+      end
+      upload(zipped)
+      exported = event_scope.update_all(payload: nil)
+      [exported, s3_key]
+    end
+
+  private
+
+    attr_reader :created_before, :created_on_or_after
+
+    def bucket
+      @bucket ||= s3.bucket(Rails.application.config.s3_export.bucket)
+    end
+
+    def s3
+      @s3 ||= Aws::S3::Resource.new(region: Rails.application.config.s3_export.region)
+    end
+
+    def build_csv(file)
+      csv = CSV.new(file)
+      event_scope.find_each.with_index do |event, index|
```

Comment:
How about...
```ruby
csv << Event.new.as_csv.keys
event_scope.find_each do |event|
  csv << event.as_csv.values
end
```

Simplifies the logic. I had to work out what the index was needed for.